Sources/Hummingbird/Files/LocalFileSystem.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import Logging
+import NIOPosix
+
+/// Local file system file provider. All file accesses are relative to a root folder
+public struct LocalFileSystem: FileProvider {
+    let rootFolder: String
+    let fileIO: FileIO
+
+    /// Initialize LocalFileSystem FileProvider
+    /// - Parameters:
+    ///   - rootFolder: Root folder to serve files from
+    ///   - threadPool: Thread pool used when loading files
+    ///   - logger: Logger to output root folder information
+    init(rootFolder: String, threadPool: NIOThreadPool, logger: Logger) {
+        if rootFolder.last != "/" {
+            self.rootFolder = "\(rootFolder)/"
+        } else {
+            self.rootFolder = rootFolder
+        }
+        self.fileIO = .init(threadPool: threadPool)
+
+        let workingFolder: String
+        if rootFolder.first == "/" {
+            workingFolder = ""
+        } else {
+            if let cwd = getcwd(nil, Int(PATH_MAX)) {
+                workingFolder = String(cString: cwd) + "/"
+                free(cwd)
+            } else {
+                workingFolder = "./"
+            }
+        }
+        logger.info("Serving files from \(workingFolder)\(rootFolder)")
+    }
+
+    /// Get full path name with local file system root prefixed
+    /// - Parameter path: path from URI
+    /// - Returns: Full path
+    public func getFullPath(_ path: String) -> String {
+        if path.first == "/" {
+            return "\(self.rootFolder)\(path.dropFirst())"
+        } else {
+            return "\(self.rootFolder)\(path)"
+        }
+    }
+
+    /// Get file attributes
+    /// - Parameter path: Full path to file
+    /// - Returns: File attributes
+    public func getAttributes(path: String) async throws -> FileAttributes? {
+        do {
+            let lstat = try await self.fileIO.fileIO.lstat(path: path)
+            let isFolder = (lstat.st_mode & S_IFMT) == S_IFDIR
+            let modificationDate = Double(lstat.st_mtimespec.tv_sec) + (Double(lstat.st_mtimespec.tv_nsec) / 1_000_000_000.0)
+            return .init(
+                isFolder: isFolder,
+                size: numericCast(lstat.st_size),
+                modificationDate: Date(timeIntervalSince1970: modificationDate)
+            )
+        } catch {
```

Comment:
Is there a known error that would be thrown if the file does not exist?

Reply:
It's a `NIOCore.IOError(2)` but this could change when we move to NIOFileSystem.